Add blog post about the new error-handling logic. (#17)
CHERIoT-Platform/CHERIoT-Platform.github.io: _posts/2024-09-20-error-handling.markdown

---
layout: post
title:  "Improved error handling in CHERIoT RTOS"
date:   2024-09-20
categories: rtos errors
author: "David Chisnall"
---

CHERI platforms in general, and CHERIoT in particular, can turn a lot of bugs that would be silent data corruption into recoverable errors.

> The 'recoverable' part comes from the fact that the error is caught *before* you perform the invalid operation.

Reviewer: ```suggestion
The 'recoverable' part comes from the fact that any error is caught *before* the system would commit an invalid operation.
```

Exceptions are a form of non-local return: they transfer control flow to a function higher up the stack to report the error.

On most modern platforms, exceptions are implemented using a *table-driven unwinder*.
The exact details differ between 64-bit Windows and *NIX systems, but the underlying mechanisms are very similar.
For each function, the compiler emits a table that describes how to unwind the stack.
A generic unwinder library can read this information and pop each frame off the call stack, one at a time.
Language-specific exceptions are built on top of this.
In the unwind data, each function can define a *personality function* that uses some language-specific data to determine whether the current unwind should run cleanups or should stop for a catch.

All of this adds up to a lot of code.
The unwind library and C++ exception support are typically on the order of 2-300 KiB of code.
On top of this, the unwind metadata can add 10-20% to the size of a program.
This means that, for a lot of embedded systems, the generic unwinder plus the unwind metadata would consume all of memory, leaving no space for anything else.

For CHERIoT, we went back to older mechanisms.

> Two systems implemented exceptions in a similar way: 16-bit (and 32-bit) Windows, and OpenStep.

Reviewer: Also Plan 9, FWIW

Pointers derived from the stack pointer and return addresses may be stored only on the stack.
This means that at least two of the four registers that must be stored in a `jmp_buf` by `setjmp` can be stored only on the stack.

The traditional approach of keeping the head of the linked list in a global is therefore impossible.
The slightly more modern approach of using thread-local storage (TLS) would also not work because we do not have thread-local storage.

The obvious solution is therefore to add TLS, but what would that mean?
Consider a single thread that starts in compartment A, then calls into compartment B, which calls back into A.
If we have conventional TLS, A could stash a pointer into the part of the stack that it passes to B in TLS and then in the second invocation it would be able to violate compartment isolation (reading and writing all of the on-stack state for B).
If we built a linked-list of `jmp_buf`s there, then it would be possible for the second invocation of A to jump back to the first, bypassing B and violating the trusted-stack guarantees.
This would be bad.

What we want is not *thread-local* storage but *compartment-invocation-local* storage.

> Each time a thread enters a new compartment, you should get some storage that is not tied to the depth in the control stack.

Reviewer: Erm?  How do you mean "not tied to the depth in the control stack"?

To implement this, we looked at the earliest way that operating systems have implemented thread-local storage: reserve some space at the top of the stack.

Now, on entry into a compartment, the switcher will move the stack pointer 16 bytes down before transferring control into the callee.
Similarly, the loader will reserve 16 bytes at the top of the stack before starting a thread.

> This means that you have two pointers worth of space that are easy to find (set the address of the stack pointer to its top, then set the address to eight or 16 bytes below that).

Reviewer: Possibly worth explicitly reminding the reader that the switcher sets bounds on csp on cross-compartment invocations?

> CHERIoT has a convenient `cgettop` instruction and so this sequence is very short.  First, `cgettop` gives the top address, then `csetaddress` gives a new capability derived from the stack pointer that points to the address.  After that, a `-8` immediate offset to a load or store capability instruction can access the space, so we need only three instructions to load the head of the list.

Reviewer: By convention this should be multiple lines.  Also hey look, a one-past pointer that we are going to bring back in bounds (as opposed to just a loop terminator or something of that ilk).

> With this, we can store the head of the linked list of error handlers at the top of the stack.

Reviewer 2: part of the stack used for the current compartment invocation.

When you want to jump to the nearest error handler, find the head of the list relative to the stack pointer, pop it, and pass it to `longjmp`.

> The `cleanup_unwind` does all of this for you, so typically you won't need to ever see that this is how it's implemented.

Reviewer: `cleanup_unwind` method? library call?

Everything so far is enough to build a set of nested error handlers, but what happens if the error is caused by running out of stack space?
[PR 301](https://github.com/CHERIoT-Platform/cheriot-rtos/pull/301) adds the last bit of this: support for a new kind of error handler that doesn't use a stack.

This is in addition to the existing error handlers.
The stackless error handler (if it exists) will run instead, in any of the following conditions:

 - A compartment doesn't provide the normal error handler.
 - There isn't enough stack space available for the context.
 - The stack isn't valid at all (the stack pointer can become untagged if a function prologue subtracts from it and moves it out of bounds).

This means that you can now run an error handler even if you overflow the stack.
But what happens if you want to jump to the nearest error handler registered as described earlier and the stack pointer isn't valid?

First, the CPU will trap because a `csp`-relative load fails because `csp` (the capability stack pointer register) is untagged.
This transitions to the switcher.
The switcher will then find that you need to run the stackless error handler (because `csp` is untagged).

> The switcher will then look at the trusted stack and rederive the `csp` value that you had on entry.

Reviewer: ```suggestion
The switcher will then look at the trusted stack and rederive the `csp` value that your compartment invocation had on entry.
```

> The stackless error handler *does* have a stack, but it doesn't have a *stack frame*.

Reviewer: ```suggestion
That is, the stackless error handler has *access to* the stack, but it doesn't have an *activation frame* on the stack.
```

When it's invoked, it can guarantee that `csp` is a capability that authorises access to the stack, but it can't guarantee anything about the address of that capability (other than that it will be in bounds).

That's absolutely fine for popping the top error handler and jumping to it though.

# Putting it all together

Most users will never need to know anything about the above.
They can simply use the macros or wrapper functions, and the error handler, that we provide.
The error handler is added to a compartment by adding the following line under the `compartment` declaration in `xmake.lua`:

```lua
    add_deps("unwind_error_handler")
```

C programmers have to use the macro-based version, which looks like this:

```c
CHERIOT_DURING
{

>     // Do some things that may cause a 

Reviewer 2: crash?

}
CHERIOT_HANDLER
{
    // Handle the failure
}
CHERIOT_END_HANDLER
```

Note that, because these are implemented with `setjmp`, the usual rules about `setjmp` apply.
Specifically, anything that's used in both blocks must be declared `volatile`.

In C++, you can use the `on_error` function, which takes two lambdas (or other callable objects), representing the try and error paths.
If you're using RAII for cleanup, you can often omit the second.
For example:

```c++
    LockGuard g(flagLock);
    // No handler.  g's destructor runs after on_error returns.
    on_error([&]() { /* code that might fail here */ });
```

This will run the code in the lambda and then release the lock, even if you take a CHERI trap.

Hopefully this is much easier than writing a custom error handler that detected held locks and released them.

This isn't a full exception model.
In particular, there's currently no way in the handler of seeing the cause of the fault.

> In a future iteration, we'll add something like 'Herbceptions': Herb Sutter's proposal for C++ where exceptions must be preallocated objects or primitive values.

Reviewer: Link?
